### PR TITLE
Remove hook metadata from DOM element on cleanup

### LINKED ISSUE #1174: Hook not getting remounted after liveview reconnect
### Environment

* Elixir version (elixir -v): Elixir 1.10.4
* Phoenix version (mix deps): 1.5.5
* Phoenix LiveView version (mix deps): 0.14.7
* NodeJS version (node -v): v14.13.0
* NPM version (npm -v): 6.14.8
* Operating system: mac
* Browsers you attempted to reproduce this bug on (the more the merrier): chrome, safari
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: yes

I have a setup where I have a parent liveview for the dashboard and then a separate liveview for each widget on the dashboard. I've noticed that when the parent liveview disconnects (and crashes?), phx-hooks defined on the dashboard (parent liveview) get `reconnected`, but the phx-hooks that are inside the child liveviews, get neither `reconnected` nor `mounted` again (only `destroyed` is called). I can prepare a repro if needed.

[Browser logs](https://gist.github.com/syfgkjasdkn/bd483a0f730b49c7e9b2e7d065e4d5a6)

### PR DESCRIPTION
**Pull request by jonatanklosko**
Fixes #1174.

When the socket is closed the, root `LiveView` recursively destroys all the nested `LiveViews`. As part of this process we remove hooks for each child:

https://github.com/phoenixframework/phoenix_live_view/blob/95d5c7ccd0ac66e04b15c7b6128d44b60767e682/assets/js/phoenix_live_view.js#L2033-L2036

Which further leads to:

https://github.com/phoenixframework/phoenix_live_view/blob/95d5c7ccd0ac66e04b15c7b6128d44b60767e682/assets/js/phoenix_live_view.js#L2380-L2384

---

The problem comes from the fact that during Hook initialization we add `phxHookId` metadata to the DOM element:

https://github.com/phoenixframework/phoenix_live_view/blob/95d5c7ccd0ac66e04b15c7b6128d44b60767e682/assets/js/phoenix_live_view.js#L2890

and we never actually remove this metadata, event when the hook is destroyed!

On reconnect, the child `LiveView`s are mounted again and get to the hook initialization:

https://github.com/phoenixframework/phoenix_live_view/blob/95d5c7ccd0ac66e04b15c7b6128d44b60767e682/assets/js/phoenix_live_view.js#L2365

but as we never removed the `phxHookId` from the DOM element, the expression `ViewHook.elementID(el)` returns the old id and new hook is not created (even though it should).

---

I believe we should remove the metadata and the `ViewHook#__cleanup__` method looks like the appropriate place.

### CODE REVIEW

File: assets/js/phoenix_live_view.js [hunk 1]
```diff
@@ -2379,8 +2379,8 @@ export class View {
 
   destroyHook(hook){
     hook.__trigger__("destroyed")
-    hook.__cleanup__()
     delete this.viewHooks[ViewHook.elementID(hook.el)]
+    hook.__cleanup__()
```
Comment (jonatanklosko) [inline assets/js/phoenix_live_view.js:2383]: Note: the call to `ViewHook.elementID(hook.el)` uses the `phxHookId`, so we need to execute that first and cleanup afterwards.

File: assets/js/phoenix_live_view.js [hunk 2]
```diff
@@ -2916,6 +2916,7 @@ class ViewHook {
 
   __cleanup__(){
     this.__listeners.forEach(callbackRef => this.removeHandleEvent(callbackRef))
+    delete this.el.phxHookId
```
Comment (mveytsman) [inline assets/js/phoenix_live_view.js:2919]: I'm noticing that this is the only time we set attributes on elements directly, and everywhere else we DOM.putPrivate/DOM.private.

I wonder if makes sense to use `DOM.putPrivate` to set the hook id, and then we can remove everything in DOM.private of the view element as part of cleanup?

My thinking is that only ever using the private attribute for metadata like this and cleaning it all up will help prevent this class of bug in the future, but maybe it's more of a refactor than we want to do for this / there's a difference in the phxHookId and other things we store in that private store...

**Comment by chrismccord** [comment]
@jonatanklosko continuing our convo from #1174  – There's a lot of context on how this works to understand the bug, so let me run through the nuances so we're all on the same page:

1. Hooks are mounted when a DOM node is added to the DOM, and *never again* while that DOM node remains in the DOM. This is important because users will wire up event listeners and other element state in `mounted` and we cannot cause double bindings.
2. When the LiveVIew process loses connection (via disconnect or LV crash), disconnected is called.
3. We only invoke `reconnected` if the DOM node remains in the DOM across a patch after connection/LV recovery.
4. In past LV versions, we autogenerated container IDs for nested LV children, but now we require an `:id`, which we set as at the DOM ID.
5. When IDs were autogenerated, we always destroyed and remounted hooks because the parent DOM node was removed from the DOM and new one was added (via the random ID)
6. Now that children have fixed IDs, we technically can only invoke `reconnected` because the DOM nodes stays in the DOM, but today we destroy the child hooks on crash/disconnect so this is the bug we are discussing

The larger issue we need to address is how to handle hook instances in the case where the DOM node remains in the DOM but we destroy our references to the hook. We may be able to hoist the hooks up to the root LV, but I need to do some more thinking on this. Going back to our random IDs may be the best route, in which case this issue may be resolved without modification. For example, in your repro, you can try something like this to see it in action:

            <%= live_render @socket, CodesWeb.ChildLive, id: "test-#{System.unique_integer([:positive])}" %>

Make sense?

**Comment by jonatanklosko** [comment]
Thanks for the context, it all makes sense and helps a lot!

> This is important because users will wire up event listeners and other element state in mounted and we cannot cause double bindings.

Theoretically they should cleanup listeners in the `destroyed` thought, right?

---

On a related note, one thing I noticed is that a hook has a direct reference to the view (`__view`), but when we destroy a child view on disconnect, then this reference is no longer valid (at least conceptually). To draw an example, consider the following `mount` method:

```js
mount() {
  this.el.addEventListener('click', event => {
    this.pushEvent('click', { value: 'cats' });
  });
}
```

Now when when the connection is lost, the view gets destroyed, but clicking would still use the `__view` reference to try pushing events to the socket.

So from this example it seems natural to me to remove a hook when the corresponding view is destroyed. Also, assuming the user properly implements the `destroy` method, there should be no downside to mounting the hook again (?).

---

Also I discovered that the change I suggested triggers `mounted`, but also `reconnected` (because the hook is added back to the list of all hooks and `reconnected` is triggered for all the hooks).

**Comment by chrismccord** [comment]
> Theoretically they should cleanup listeners in the `destroyed` thought, right?

One of the features of hooks is exactly to have to cleanup DOM related aspects. The issues is DOM nodes previously did not survive beyond their parent LiveView, because the LV element was removed from the DOM (new ID).



> Now when when the connection is lost, the view gets destroyed, but clicking would still use the `__view` reference to try pushing events to the socket.

This would be an issue, but per the point above, shouldn't actually happen under the model where a LV owns the DOM nodes and they live and die with the LV itself. With the bug today that is not aways the case, but to be clear the hook model is 1:1 lifecycle with the LiveView.

Give me some time today to dive in as this requires some deeper changes to properly fix if we want to maintain the current hook model. I think we'll need to avoid destroying the LiveView child instances if they are to be recovered and have them act similarly to how the root LiveView is handled, but I need to really dig in to be sure. Stand by! Thanks!

**Comment by jonatanklosko** [comment]
> I think we'll need to avoid destroying the LiveView child instances if they are to be recovered

Sounds like a good approach to look into. I was wondering if there's a specific reason why they need to be destroyed, given they are kinda standalone anyway.

Thanks, and if there's anything I can help with let me know =)

**Comment by chrismccord** [comment]
Sorry typo above: One of the features of hooks is exactly to avoid* having to cleanup DOM related aspects.

**Comment by chrismccord** [comment]
Destroying was more of a convenience because the internal channel lifecycle needs to be joined in a specific order across the children, as well as the fact that children may actually never be added back to the page after rejoin patch.

**Comment by chrismccord** [comment]
I solved this on the morphdom side by using `getNodeKey` to uniquely identify the LV containers by ID + session, which will teardown and recreate the trees when we patch with a new session, such as child recovery. Longer term, I think we can generalize this with a morph attr signifying which ID to use, that way the complexity can be on the server rather than client, but this is fine for now. Thanks for the help on this! I really appreciate it.

**Comment by jonatanklosko** [comment]
That's a clever solution, great!

I believe there's one tiny detail left, on reconnect the newly created hooks have their two methods called: `mounted` then `reconnected` (that's because we first [add the new hook](https://github.com/phoenixframework/phoenix_live_view/blob/f3e678c6356e440ae16bed7557a00c2ae3e0bb95/assets/js/phoenix_live_view.js#L2187-L2190) and then [call reconnect for all](https://github.com/phoenixframework/phoenix_live_view/blob/f3e678c6356e440ae16bed7557a00c2ae3e0bb95/assets/js/phoenix_live_view.js#L2201)). Is this desired for newly created hooks? (it could be surprising call `reconnected` without calling `disconnected` prevoiusly)

**Comment by chrismccord** [comment]
That was the previous behavior but you're right that reconnected only makes sense if we had previously disconnected on the hook. Do you want to take a stab at only invoking reconnected on previously disconnected hooks?

**Comment by jonatanklosko** [comment]
Sure, gonna send a PR shortly!

**Pull request closed without merging**